storage/remote/queue_manager.go

```diff
@@ -504,58 +514,115 @@ func (t *QueueManager) sendMetadataWithBackoff(ctx context.Context, metadata []p
 	return nil
 }
 
-// Append queues a sample to be sent to the remote storage. Blocks until all samples are
+// Append takes a set of samples to be sent to the remote storage, and
+// breaks them into sets to be sent to each shard. Blocks until all samples are
 // enqueued on their shards or a shutdown signal is received.
 func (t *QueueManager) Append(samples []record.RefSample) bool {
-	var appendSample prompb.Sample
-outer:
+	var toSend []sampleOrExemplar
+	pooledBuf := t.sendPool.Get()
+	if pooledBuf != nil && cap(pooledBuf.([]sampleOrExemplar)) <= len(samples) {
+		toSend = pooledBuf.([]sampleOrExemplar)[:0]
+	} else {
+		toSend = make([]sampleOrExemplar, 0, len(samples))
+	}
+	defer t.sendPool.Put(toSend)
+	// Turn the record.RefSample objects that we receive into sampleOrExemplar
+	// objects that have a reference to the series labels.
+	t.seriesMtx.Lock()
 	for _, s := range samples {
-		t.seriesMtx.Lock()
 		lbls, ok := t.seriesLabels[s.Ref]
 		if !ok {
 			t.metrics.droppedSamplesTotal.Inc()
 			t.dataDropped.incr(1)
 			if _, ok := t.droppedSeries[s.Ref]; !ok {
 				level.Info(t.logger).Log("msg", "Dropped sample for series that was not explicitly dropped via relabelling", "ref", s.Ref)
 			}
-			t.seriesMtx.Unlock()
 			continue
 		}
-		t.seriesMtx.Unlock()
-		// This will only loop if the queues are being resharded.
-		backoff := t.cfg.MinBackoff
-		for {
+		toSend = append(toSend, sampleOrExemplar{
+			ref:          s.Ref,
+			seriesLabels: lbls,
+			timestamp:    s.T,
+			value:        s.V,
+			isSample:     true,
+		})
+	}
+	t.seriesMtx.Unlock()
+	return t.enqueue(toSend)
+}
+
+func (t *QueueManager) enqueue(toSend []sampleOrExemplar) bool {
+	backoff := t.cfg.MinBackoff
+	// We come back here with un-sent data in toSend if the QueueManager reshards,
+	// or if there were so many entries in toSend that we reached capacity on one batch.
+	for len(toSend) > 0 {
+		nShards := t.shards.numCurrentShards()
+		// Distribute toSend across nShards batches of max size Capacity.
+		batches := make([][]sampleOrExemplar, nShards)
+		index := 0
+		for ; index < len(toSend); index++ {
+			s := toSend[index]
+			shard := uint64(s.ref) % uint64(nShards)
+			if batches[shard] == nil {
+				batches[shard] = t.batchPool.Get().([]sampleOrExemplar)[:0]
+			}
+			batches[shard] = append(batches[shard], s)
+			// Once one shard's batch reaches capacity, break this loop and go to sending phase
+			if len(batches[shard]) >= t.cfg.Capacity {
+				index++
+				break
+			}
+		}
+		toSend = toSend[index:]
+
+		// For each batch, attempt to enqueue on the shard.
+		for shard, data := range batches {
+			if len(data) == 0 {
+				continue
+			}
 			select {
 			case <-t.quit:
 				return false
 			default:
 			}
-			appendSample.Value = s.V
-			appendSample.Timestamp = s.T
-			if t.shards.enqueue(s.Ref, writeSample{lbls, appendSample}) {
-				continue outer
+			if t.shards.enqueue(shard, nShards, data) {
```

Comment:
Since the underlying channel only has a capacity of 1 now, would this have the potential to block sending to any shard well before the capacity of a single shard is filled?

Reply:
Good point.  Can't see a way forward except by abandoning the current meaning of `Capacity`.
(or your PR)